Mathlib/Analysis/Normed/ValuativeRel.lean

```diff
@@ -0,0 +1,101 @@
+/-
+Copyright (c) 2025 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import Mathlib.Analysis.Normed.Field.Lemmas
+import Mathlib.Analysis.Normed.Group.Ultra
+import Mathlib.RingTheory.Valuation.RankOne
+import Mathlib.Topology.Algebra.Valued.ValuativeRel
+
+/-!
+
+# Nonarchimedean normed rings carry a valuative topology
+
+In this file, we provide a helper instance
+for `ValuativeRel R` derived from a `NormedRing R` when `NormMulClass R` and `IsUltrametricDist R`,
+so that downstream files can refer to `ValuativeRel R`,
+to facilitate a refactor.
+
+## Implementation details
+
+The namespace is `NormedField` because we need `‖x * y‖ = ‖x‖ * ‖y‖`, not just submultiplicativity.
+The `NormedCommRing` assumption is not enough because that gives `‖x * y‖ ≤ ‖x‖ * ‖y‖`.
+The associated prop classes of `NormOneClass` and `NormMulClass` give the full multiplicativity.
+
+-/
+
+variable {R : Type*} [NormedCommRing R] [NormOneClass R]
+  [NormMulClass R] [IsUltrametricDist R]
+
+open Filter Set Valuation
+
+open scoped NNReal
+
+namespace NormedField
+
+/-- The valuation on a nonarchimedean normed field `K` defined as `nnnorm`.
+Should replace `NormedField.valuation` in the future. -/
+def valuation' : Valuation R ℝ≥0 where
+  toFun           := nnnorm
+  map_zero'       := nnnorm_zero
+  map_one'        := nnnorm_one
+  map_mul'        := nnnorm_mul
+  map_add_le_max' := IsUltrametricDist.norm_add_le_max
+
+@[simp]
+theorem valuation'_apply (x : R) : valuation' x = ‖x‖₊ := rfl
+
+/-- A nonarchimedean normed ring carries a valuative relation induced by the norm.
+This is a scoped instance. -/
+def toValuativeRel : ValuativeRel R := .ofValuation valuation'
+
+open Filter Topology
+
+scoped[NormedField] attribute [instance] NormedField.toValuativeRel
+
+lemma rel_iff_le (x y : R) : x ≤ᵥ y ↔ ‖x‖ ≤ ‖y‖ := Iff.rfl
+
+lemma norm_pos_posSubmonoid (x : (ValuativeRel.posSubmonoid R)) :
+    0 < ‖(x : R)‖ := by
+  simp
+
+instance : (valuation' (R := R)).Compatible := .ofValuation _
+
+instance : ValuativeRel.IsRankLeOne R :=
+  .of_compatible_mulArchimedean (valuation' (R := R))
+
+instance {K : Type*} [NormedField K] [IsUltrametricDist K] :
```

Comment:
If `NormedField.toValuativeRel` is scoped, should all these be scoped as well?

Reply:
It seems that it doesn't make a difference, because they won't fire without the scope opened:
```
variable {K : Type*} [NormedField K] [IsUltrametricDist K]

/-- error: failed to synthesize
  ValuativeRel K

Hint: Additional diagnostic information may be available using the `set_option diagnostics true` command.  -/
#guard_msgs in
#synth Valuation.Compatible (NormedField.valuation' (R := K))
```

Reply:
But I am guessing that they would be tried and potentially cause bad performance.